snowfakery/standard_plugins/Salesforce.py

```diff
@@ -7,8 +17,150 @@
+from snowfakery.parse_recipe_yaml import TableInfo
+
+
+MAX_SALESFORCE_OFFSET = 2000  # Any way around this?
+
+# the option name that the user specifies on the CLI or API is just "orgname"
+# but using this long name internally prevents us from clashing with the
+# user's variable names.
+plugin_option_name = "snowfakery.standard_plugins.Salesforce.SalesforceQuery.orgname"
+
+
+class SalesforceConnection:
+    """Helper layer above simple_salesforce and salesforce_bulk"""
+
+    _sf = None
+
+    def __init__(self, get_orgname):
+        self.get_orgname = get_orgname
+        self.logger = getLogger(__name__)
+
+    @property
+    def sf(self):
+        """simple_salesfore client"""
+        if not self._sf:
+            self._sf, self._bulk = self._get_sf_clients(self.orgname)
+        return self._sf
+
+    def bulk(self):
+        """salesforce_bulk client"""
+        if not self._bulk:
+            self._sf, self._bulk = self._get_sf_clients(self.orgname)
+        return self._bulk
+
+    @property
+    def orgname(self):
+        """Look up the orgname in the scope"""
+        return self.get_orgname()
+
+    def query(self, *args, **kwargs):
+        """Query Salesforce through simple_salesforce"""
+        return self.sf.query(*args, **kwargs)
+
+    def query_single_record(self, query):
+        """Query Salesforce through simple_salesforce and
+        validate that query returns 1 and only 1 record"""
+        qr = self.sf.query(query)
+        records = qr.get("records")
+        if not records:
+            raise DataGenValueError(f"No records returned by query {query}", None, None)
+        elif len(records) > 1:
+            raise DataGenValueError(
+                f"Multiple records returned by query {query}", None, None
+            )
+        record = records[0]
+        if "attributes" in record:
+            del record["attributes"]
+        if len(record.keys()) == 1:
+            return tuple(record.values())[0]
+        else:
+            return PluginResult(record)
+
+    def compose_query(self, context_name, **kwargs):
+        kwargs = kwargs.copy()
+        fields = kwargs.pop("fields", None)
+        sobject = kwargs.pop("from", None)
+        where = kwargs.pop("where", None)
+
+        if not fields:
+            raise DataGenError(f"{context_name} needs a 'fields' list")
+
+        if not sobject:
+            raise DataGenError(f"{context_name} needs a 'from'")
+
+        if kwargs:
+            raise DataGenError(
+                f"Unknown argument in {context_name}: {tuple(kwargs.keys())}"
+            )
+
+        query = f"SELECT {fields} FROM {sobject} "
+        if where:
+            query += f" WHERE {where}"
+        return query
+
+    @staticmethod
+    def _get_sf_clients(orgname):
+        try:
+            runtime = CliRuntime(load_keychain=True)
+        except Exception as e:
+            raise DataGenError("CLI Runtime cannot be loaded", *e.args)
```

Comment:
This looks odd; why would the args for an arbitrary exception from CliRuntime be a filename and line_num for DataGenError?